### PR TITLE
[utils] imporove parse_duration to handle more formats

### PR DESCRIPTION
**Pull request by remitamine**
related to the previous work on dash support https://github.com/rg3/youtube-dl/pull/8408#issuecomment-179783768.

### CODE REVIEW

File: youtube_dl/utils.py [hunk 1]
```diff
@@ -1540,44 +1540,35 @@ def parse_duration(s):
...
-        res += int(m.group('days')) * 24 * 60 * 60
-    if m.group('ms'):
-        res += float(m.group('ms'))
-    return res
+    days, hours, mins, secs, ms = None, None, None, None, None
+    m = re.match(r'(?:(?:(?:(?P<days>[0-9]+):)?(?P<hours>[0-9]+):)?(?P<mins>[0-9]+):)?(?P<secs>[0-9]+)(?P<ms>\.[0-9]+)?$', s)
+    if m:
+        days, hours, mins, secs, ms = m.groups()
+    else:
+        m = re.match(r'(?i)(?:P?T)?(?:(?P<days>[0-9]+)\s*d(?:ays?)?\s*)?(?:(?P<hours>[0-9]+)\s*h(?:ours?)?\s*)?(?:(?P<mins>[0-9]+)\s*m(?:in(?:ute)?s?)?\s*)?(?:(?P<secs>[0-9]+)(?P<ms>\.[0-9]+)?\s*s(?:ec(?:ond)?s?)?\s*)?$', s)
```
Comment (dstftw) [inline youtube_dl/utils.py]: This regex should be split into multiple lines for bettercode perception.

File: youtube_dl/utils.py [hunk 2]
```diff
@@ -1540,44 +1540,35 @@ def parse_duration(s):
...
+    if mins:
+        duration += float(mins) * 60
+    if hours:
+        duration += float(hours) * 60 * 60
+    if days:
+        duration += float(days) * 24 * 60 * 60
+    if ms:
+        duration += float(ms)
+    if duration.is_integer():
+        duration = int(duration)
```
Comment (dstftw) [inline youtube_dl/utils.py]: I guess it's ok to always return stable float here.

File: youtube_dl/utils.py (hunk 1 above)
Comment (remitamine) [inline youtube_dl/utils.py]: like that?
```python
        m = re.match(r'''(?ix)(?:P?T)?
            (?:(?P<days>[0-9]+)\s*d(?:ays?)?\s*)?
            (?:(?P<hours>[0-9]+)\s*h(?:ours?)?\s*)?
            (?:(?P<mins>[0-9]+)\s*m(?:in(?:ute)?s?)?\s*)?
            (?:(?P<secs>[0-9]+)(?P<ms>\.[0-9]+)?\s*s(?:ec(?:ond)?s?)?\s*)?$''', s)
```

Comment (dstftw) [inline youtube_dl/utils.py]: I would prefer indent similar to the former code.

File: youtube_dl/utils.py [hunk 3]
```diff
@@ -1540,44 +1540,35 @@ def parse_duration(s):
...
-    if m.group('hours'):
-        res += int(m.group('hours')) * 60 * 60
-    if m.group('hours_reversed'):
-        res += int(m.group('hours_reversed')) * 60 * 60
-    if m.group('days'):
-        res += int(m.group('days')) * 24 * 60 * 60
-    if m.group('ms'):
-        res += float(m.group('ms'))
-    return res
+    days, hours, mins, secs, ms = None, None, None, None, None
```
Comment (dstftw) [inline youtube_dl/utils.py]: This can be simplified to `[None] * 5` or `(None, ) * 5`.

File: youtube_dl/utils.py (hunk 1 above)
Comment (remitamine) [inline youtube_dl/utils.py]: i don't how exactly you're expect it to look like(it not the same regex).
```
        m = re.match(
            r'''(?i)(?:P?T)?
                (?:
                    (?P<days>[0-9]+)\s*d(?:ays?)?\s*
                )?
                (?:
                    (?P<hours>[0-9]+)\s*h(?:ours?)?\s*
                )?
                (?:
                    (?P<mins>[0-9]+)\s*m(?:in(?:ute)?s?)?\s*
                )?
                (?:
                    (?P<secs>[0-9]+)(?P<ms>\.[0-9]+)?\s*s(?:ec(?:ond)?s?)?\s*
                )?$''', s)
```
or this:
```
        m = re.match(
            r'''(?i)(?:P?T)?
                (?:
                    (?P<days>[0-9]+)\s*d
                    (?:ays?)?\s*
                )?
                (?:
                    (?P<hours>[0-9]+)\s*h
                    (?:ours?)?\s*
                )?
                (?:
                    (?P<mins>[0-9]+)\s*m
                    (?:in(?:ute)?s?)?\s*
                )?
                (?:
                    (?P<secs>[0-9]+)
                    (?P<ms>\.[0-9]+)?\s*s
                    (?:ec(?:ond)?s?)?\s*
                )?$''', s)
```
or something else.

Comment (dstftw) [inline youtube_dl/utils.py]: 1 is ok.

**Comment by remitamine** [comment]
all done.

**Comment by dstftw** [comment]
Thanks.

**Pull request merged**